### PR TITLE
Add `formats.javaProperties`

### PR DESCRIPTION
**Pull request by roberth**
###### Motivation

`.properties` is a deceivingly complicated format.

This PR adds a format with a quick generator that should cover most corner cases.

It also wires the `pkgs-lib` tests into `pkgs.tests` for ofborg to pick them up.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by fzakaria — APPROVED** [review]
Good tests.
LGTM -- maybe missing docs for adding the new format.

File: pkgs/pkgs-lib/formats/java-properties/default.nix [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+          #
+          # 2. iconv
+          #    Escape >ascii code points to java escapes,
+          #    as .properties files are supposed to be
+          #    encoded in ISO 8859-1. It's an old format.
+          #    UTF-8 behavior may exist in some apps and
+          #    libraries, but we can't rely on this in
+          #    general.
+
+          passAsFile = [ "value" ];
```
[the remark this answers is not in the document]
Comment (roberth) [reply pkgs/pkgs-lib/formats/java-properties/default.nix:41]: I've looked for tools, but didn't find a suitable one. For example, remarshal focuses on hierarchical semi-structured data, whereas properties are "just" a string map.
If someone knows a good, reasonably quick tool, I'll be happy to merge a replacement. I believe reasonably quick rules out the JVM.

File: pkgs/pkgs-lib/formats/java-properties/default.nix [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+          #    but may produce garbage for other control
+          #    characters.
+          #
+          # 2. iconv
+          #    Escape >ascii code points to java escapes,
+          #    as .properties files are supposed to be
+          #    encoded in ISO 8859-1. It's an old format.
+          #    UTF-8 behavior may exist in some apps and
+          #    libraries, but we can't rely on this in
+          #    general.
```
[the remark this answers is not in the document]
Comment (roberth) [reply pkgs/pkgs-lib/formats/java-properties/default.nix:39]: Thank you.
It's deceptive. A native Nix implementation (as in lib) would have a hard time performing the escaping.
A naive implementation will burden the user with escaping in their corner cases - aka the papercut experience.

File: pkgs/pkgs-lib/formats/java-properties/default.nix [hunk 3]
```diff
@@ -0,0 +1,89 @@
+{ lib, pkgs }:
+{
+  javaProperties = { comment ? "Generated with Nix" }: {
+    type = lib.types.attrsOf lib.types.str;
+
+    generate = name: value:
+      pkgs.runCommand name
```
Comment (Profpatsch) [inline pkgs/pkgs-lib/formats/java-properties/default.nix:7]: `runCommandLocal`, I don’t think substitution makes this faster.

**Review by Profpatsch — APPROVED** [review]
Pretty cool test suite. I don’t see any blockers for merging.

**Comment by roberth** [comment]
Thanks for the reviews!

**Pull request merged**